### PR TITLE
Enabled Slack Chat-bots to Respond in Threads

### PR DESCRIPTION
**Pull request by MinuraPunchihewa**
## Description

This PR enables Slack chat-bots to respond directly within a thread when a user posts a message in that thread. 

In order to achieve this, a new tables called 'threads' has been introduced to the Slack handler, which will handle all operations related to threads.

A few changes were also required to be made to the chat interfaces to allow multiple chat ID columns and multiple tables to be supported.

Fixes https://linear.app/mindsdb/issue/BE-514/slack-threads-support

**There's a slight limitation here in how the bot handles the memory (previous messages) from the thread until it is first mentioned.** @ZoranPandovski Shall I handle this in another PR?

## Type of change

- [X] ⚡ New feature (non-breaking change which adds functionality)

## Verification Process

To ensure the changes are working as expected:

 - [ ]   Test Location: Specify the URL or path for testing.
 - [ ]   Verification Steps: Outline the steps or queries needed to validate the change. Include any data, configurations, or actions required to reproduce or see the new functionality.

## Additional Media:

- [X] I have attached a brief loom video or screenshots showcasing the new functionality or change.

https://www.loom.com/share/e11f6ce62b3e4d639f71476f0c609270?sid=6fcc7827-8be9-43c9-ba14-53f6aa31c116

## Checklist:

- [X] My code follows the style guidelines(PEP 8) of MindsDB.
- [X] I have appropriately commented on my code, especially in complex areas.
- [ ] Necessary documentation updates are either made or tracked in issues - N/A
- [ ] Relevant unit and integration tests are updated or added - N/A

### CODE REVIEW

**Review by ZoranPandovski — APPROVED** [review]

File: mindsdb/interfaces/chatbot/memory.py [hunk 1]
```diff
@@ -56,9 +56,14 @@ def get_mode(self, chat_id):
...
 
-    def add_to_history(self, chat_id, chat_message):
+    def add_to_history(self, chat_id, chat_message, table_name=None):
 
-        self._add_to_history(chat_id, chat_message)
+        # If the chat_id is a tuple, convert it to a string when storing the message in the database.
+        self._add_to_history(
+            str(chat_id) if isinstance(chat_id, tuple) else chat_id,
+            chat_message,
+            table_name=table_name
```
[the remark this answers is not in the document]
Comment (MinuraPunchihewa) [reply mindsdb/interfaces/chatbot/memory.py:65]: Done! I've also made a small change to `get_chat_history()` in how `_get_chat_history()` is being called based on this.

File: mindsdb/integrations/handlers/slack_handler/slack_handler.py [hunk 2]
```diff
@@ -450,13 +570,20 @@ def _process_websocket_message(client: SocketModeClient, request: SocketModeRequ
...
                 'user': payload_event['user'],
                 'channel_id': payload_event['channel'],
                 'created_at': dt.datetime.fromtimestamp(float(payload_event['ts'])).strftime('%Y-%m-%d %H:%M:%S')
             }
 
+            # Add thread_ts to the key and row if it is a thread message. This is used to identify threads.
+            # This message should be handled via the threads table.
+            if 'thread_ts' in payload_event:
+                key['thread_ts'] = payload_event['thread_ts']
+                row['thread_ts'] = payload_event['thread_ts']
```
[the remark this answers is not in the document]
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/slack_handler/slack_handler.py]: You are right. I've removed this and also the line where `channel_id` is unnecessarily being added to `row`.

File: mindsdb/integrations/handlers/slack_handler/slack_handler.py [hunk 3]
```diff
@@ -349,6 +349,114 @@ def delete(self, query: ASTNode):
...
+            op = condition.op
+
+            # Handle the column 'channel_id'.
+            if condition.column == 'channel_id':
+                if op != FilterOperator.EQUAL:
+                    raise ValueError(f"Unsupported operator '{op}' for column 'channel_id'")
+
+                # Check if the channel exists.
+                try:
+                    channel = self.handler.get_channel(value)
```
[the remark this answers is not in the document]
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/slack_handler/slack_handler.py:384]: You are right, `conversations_replies()` does indeed handle this. The only other thing that the result of this check is being used for is to add the channel name to the response (line 415). Do you think this is unnecessary?

File: mindsdb/interfaces/chatbot/polling.py [hunk 4]
```diff
@@ -146,24 +146,42 @@ def _callback(self, row, key):
...
+                if sorted(event_keys) == sorted(table_keys):
+                    t_params = param["chat_table"]
+                    break
+
+        # Otherwise, only a single table is supported. Use the first set of parameters.
+        else:
+            t_params = self.params[0]
+
+        # Get the chat ID from the row based on the chat ID column(s).
+        chat_id = tuple(row[key] for key in t_params["chat_id_col"]) if isinstance(t_params["chat_id_col"], list) else row[t_params["chat_id_col"]]
```
[the remark this answers is not in the document]
Comment (MinuraPunchihewa) [reply mindsdb/interfaces/chatbot/polling.py:165]: It should definitely work and I agree that this is simpler. However, I believe I made a few changes based on tuple chat IDs to `MessageCountPolling` as well. Here is an [example](https://github.com/mindsdb/mindsdb/blob/4bff9eb2925ad9ead49417e3b2f30004431ae70d/mindsdb/interfaces/chatbot/polling.py#L115). Would it be OK if I took this up in another PR?

File: mindsdb/integrations/handlers/slack_handler/slack_handler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/integrations/handlers/slack_handler/slack_handler.py:384]: I haven't noticed this var is used for getting name. Ok, let it be kept as is

File: mindsdb/interfaces/chatbot/polling.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ea-rus) [reply mindsdb/interfaces/chatbot/polling.py:165]: yes, it will be ok. I thought chat_id as tuple was introduced in this PR

**Pull request merged**